src/garage/torch/algos/maml/maml.py

```diff
@@ -0,0 +1,478 @@
+                path['rewards'], self._inner_algo.discount)
+
+        self._baseline.fit(paths)
+        obs, actions, rewards, valids, baselines \
+            = self._inner_algo.process_samples(itr, paths)
+        return MAMLTrajectoryBatch(obs, actions, rewards, valids, baselines)
+
+    def evaluate_performance(self, itr, all_samples, loss_before, loss_after,
+                             kl_before, kl, policy_entropy):
+        """Evaluate performance of this batch.
+
+        Args:
+            itr (int): Iteration number.
+            all_samples (list[list[MAMLTrajectoryBatch]]): Two
+                dimensional list of MAMLTrajectoryBatch of size
+                [meta_batch_size * (num_grad_updates + 1)]
+            loss_before (float): Loss before optimization step.
+            loss_after (float): Loss after optimization step.
+            kl_before (float): KL divergence before optimization step.
+            kl (float): KL divergence after optimization step.
+            policy_entropy (float): Policy entropy.
+
+        Returns:
+            float: The average return in last epoch cycle.
+
+        """
+        tabular.record('Iteration', itr)
+
+        for i in range(self._num_grad_updates + 1):
+            all_rewards = [
+                path_rewards for task_samples in all_samples
+                for path_rewards in task_samples[i].rewards.numpy()
+            ]
+
+            discounted_returns = [
+                tensor_utils.discount_cumsum(path_rewards,
+                                             self._inner_algo.discount)[0]
+                for path_rewards in all_rewards
+            ]
+            undiscounted_returns = np.sum(all_rewards, axis=-1)
+            average_return = np.mean(undiscounted_returns)
+            self._episode_reward_mean[i].extend(undiscounted_returns)
+
+            with tabular.prefix('Update_{0}/'.format(i)):
+                tabular.record('AverageDiscountedReturn',
+                               np.mean(discounted_returns))
+                tabular.record('AverageReturn', average_return)
+                tabular.record('Extras/EpisodeRewardMean',
+                               np.mean(self._episode_reward_mean[i]))
+                tabular.record('StdReturn', np.std(undiscounted_returns))
+                tabular.record('MaxReturn', np.max(undiscounted_returns))
+                tabular.record('MinReturn', np.min(undiscounted_returns))
+                tabular.record('NumTrajs', len(all_rewards))
+
+        with tabular.prefix(self._policy.name + '/'):
+            tabular.record('LossBefore', loss_before)
+            tabular.record('LossAfter', loss_after)
+            tabular.record('dLoss', loss_before - loss_after)
+            tabular.record('KLBefore', kl_before)
+            tabular.record('KLAfter', kl)
+            tabular.record('Entropy', policy_entropy)
+
+        return average_return
+
+
+class MAMLTrajectoryBatch(
```

Comment:
i still don't see the advantage of this dtype over TrajectoryBatch. is there a reason you can't use TrajectoryBatch?

Reply:
I can investigate more time on refactoring to use `TrajectoryBatch` in the algorithm. I just feel that since we are not using samplers that return `TrajectoryBatch`, we are not solving the issue from its root cause.